flow/machine.go

```diff
@@ -0,0 +1,121 @@
+package flow
+
+import (
+	"errors"
+	"fmt"
+)
+
+// Machine describes the contract for the flow handling
+type Machine interface {
```

Comment:
I don't see how this interface would correspond to actual FSM backend... ideally, we want some kind of interface that would be easy to implement, as it may vary based on whether if you want to use cache, what kind of it, if you want to use code-generation, and so on. Consider https://github.com/d5/go-fsm for example of how FSM may be pushed quite considerably based on application.

Reply:
Yes, I've thought about that, but ultimately, my idea is very simple: you only need to call one function
for each step (that interacts with the user, I mean). But you have good thoughts regarding caching and so on.
With further consideration, we can explore alternative approaches:

```go
flow.New().
  Step(sendUserMessage("Enter email:")).
  Step(validateEnteredEmail).
  Step(assignMailToVariable).
  Step(sendUserMessage("Enter password:")).
```

Based on the code above, we can define any steps we want. Each step should return an error, indicating that the user did not successfully pass the step, and we need to wait for the next prompt and rerun the step. It's a very simple interface to implement.
By the way, we have some questions to discuss:

- Do we need to provide common handlers for the entire flow's completion? Such as fail/success handlers as implemented in my idea here.
- Do we need to have an opportunity to mark steps as internal/external? Internal means that the step doesn't interact with the user (just simple logging, for example).
- How can we solve the following problem: we want to log every step using only one function. For that, we need to provide an interface similar to the one I have now.

That's why I'm not sure if the ftm is still a good idea. But I agree that it is easier for the user and looks clearer.

Reply:
I'm quite certain we need to address that functionality, even if the user can perform validation/logging, etc., within the step function. It's unclear, and I suspect no one would want to do it. Additionally, it doesn't resolve the issue with global handlers (such as fail/logging, etc.).